### ISSUE TITLE
Segmentation fault in vec_unique_loc

### ISSUE THREAD

**Issue by MihaiBabiac** [r-lib/vctrs#1272]
Hi! I am using dplyr to preprocess a data set and the very first step is to remove all duplicates. This involves going through ~100 tibbles, each with ~1.5M rows and 21 columns. Very often (but not always) after processing some of the files, R encounters a segmentation fault and crashes. This always happens inside `vec_unique_loc`, which has been called by dplyr's `distinct` function.
```
 *** caught segfault ***
address 0x50cc8, cause 'memory not mapped'

Traceback:
 1: vec_unique_loc(as_tibble(out)[prep$vars])
 2: distinct.data.frame(df)
 3: distinct(df)
 4: ...
```
The top of the backtrace by opening the core dump in gdb:
```
#0  0x00007fd0285e37d8 in int_equal_scalar_na_equal (x=0x10b03c, y=0x10b084) at equal.h:25
#1  equal_scalar_na_equal_p (proxy_type=vctrs_type_logical, x=0x55636a4764b0, x_p=0xe0, i=i@entry=273367, y=<optimized out>, y_p=0xe0, j=273385) at equal.c:57
#2  0x00007fd0285e0c64 in df_equal (x=<optimized out>, i=273367, y=<optimized out>, j=273385) at dictionary.c:150
#3  0x00007fd0285e171e in dict_hash_with (d=d@entry=0x556379e98838, x=x@entry=0x556379e98838, i=i@entry=273385) at dictionary.c:337
#4  0x00007fd0285e18c2 in dict_hash_scalar (i=273385, d=0x556379e98838) at dictionary.c:381
#5  vctrs_unique_loc (x=<optimized out>) at dictionary.c:381
```

I event tried to make a minimal crashing example, but without luck. It is not a consistent bug which might make it harder to pinpoint. If it helps, I can also provide you with the complete core dump.
System information: vctrs 0.3.1, dplyr 1.0.0, R 4.0.2, Ubuntu 20.04.1

Thanks for your help!

**Comment by DavisVaughan** [comment]
Can you update vctrs? It looks like you have 0.3.1, and we fixed an issue like this in 0.3.2 https://github.com/r-lib/vctrs/pull/1167. FWIW the CRAN version is now 0.3.4

**Comment by MihaiBabiac** [comment]
That seems to have worked. Thanks a lot!
